core/src/main/scala/com/avast/grpc/jsonbridge/ReflectionGrpcJsonBridge.scala

```diff
@@ -0,0 +1,151 @@
+  }
+  private val inProcessChannel = InProcessChannelBuilder.forName(inProcessServiceName).directExecutor().build()
+
+  override def close(): Unit = {
+    inProcessChannel.shutdownNow()
+    inProcessServer.shutdownNow()
+    ()
+  }
+
+  // JSON body and headers to a response (fail status or JSON response)
+  protected type HandlerFunc = (String, Map[String, String]) => F[Either[Status, String]]
+
+  // map from full method name to a function that invokes that method
+  protected val handlersPerMethod: Map[String, HandlerFunc] =
+    inProcessServer.getImmutableServices.asScala
+      .flatMap(sd => {
+        val generatedClass = getServiceGeneratedClass(sd.getServiceDescriptor)
+        val futureStub = JavaGenericHelper.asAbstractStub(
+          generatedClass
+            .getDeclaredMethod("newFutureStub", classOf[Channel])
+            .invoke(null, inProcessChannel))
+        val methods = sd.getMethods.asScala
+          .filter(supportedMethod)
+          .map(m => {
+            val requestMarshaller = m.getMethodDescriptor.getRequestMarshaller.asInstanceOf[PrototypeMarshaller[_]]
+            val requestMessagePrototype = requestMarshaller.getMessagePrototype.asInstanceOf[Message]
+            val methodName = m.getMethodDescriptor.getFullMethodName.split('/')(1)
+            val javaMethodName = methodName.substring(0, 1).toLowerCase + methodName.substring(1)
+            val stubMethod = futureStub.getClass.getDeclaredMethod(javaMethodName, requestMessagePrototype.getClass)
+            val handler: HandlerFunc = (json, headers) =>
+              Task
+                .deferFuture {
+                  val md = new Metadata()
+                  headers.foreach { case (k, v) => md.put(Metadata.Key.of(k, Metadata.ASCII_STRING_MARSHALLER), v) }
+                  val stubWithHeaders: Any = JavaGenericHelper.attachHeaders(futureStub, md)
+                  val requestBuilder = requestMessagePrototype.newBuilderForType()
+                  val request: Either[Status, Message] = try {
+                    parser.merge(json, requestBuilder)
+                    Right(requestBuilder.build())
+                  } catch {
+                    case NonFatal(ex) =>
+                      val message = "Error while converting JSON to GPB"
+                      logger.warn(message, ex)
+                      ex match {
+                        case e: StatusRuntimeException =>
+                          Left(richStatus(e.getStatus, message, e.getStatus.getCause))
+                        case _ =>
+                          Left(richStatus(Status.INVALID_ARGUMENT, message, ex))
+                      }
+                  }
+                  request match {
+                    case Right(r) =>
+                      val listenableFuture = stubMethod.invoke(stubWithHeaders, r).asInstanceOf[ListenableFuture[MessageOrBuilder]]
+                      toFuture(listenableFuture).map(response => Right(printer.print(response)))
```

Comment:
This is just ugly. If only you kept the extension method...

Reply:
This is actually more readable for me because it doesn't use implicits.